src/scatter.jl

```diff
@@ -147,3 +147,127 @@ function scatter(op::typeof(mean),
+    @eval function rrule(::typeof(scatter!), op::typeof($op), dst::AbstractArray, src::AbstractArray, idx::AbstractArray, dims)
+        y = scatter!(op, copy(dst), src, idx, dims)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_dst!(op, Δ), ∇scatter_src!(op, Δ, idx), DoesNotExist(), DoesNotExist())
+        y, $pullback
+    end
+
+    pullback = Symbol(:scatter_, opname(op), :_pullback)
+    @eval function rrule(::typeof(scatter), op::typeof($op), src::AbstractArray, idx::AbstractArray)
+        y = scatter(op, src, idx)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_src(op, Δ, idx), DoesNotExist())
+        y, $pullback
+    end
+end
+
+for op in [*, /]
+    pullback = Symbol(:scatter!_, opname(op), :_pullback)
+    @eval function rrule(::typeof(scatter!), op::typeof($op), dst::AbstractArray, src::AbstractArray, idx::AbstractArray, dims)
+        y = scatter!(op, copy(dst), src, idx, dims)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_dst!(op, Δ), ∇scatter_src!(op, Δ, dst, src, idx), DoesNotExist(), DoesNotExist())
+        y, $pullback
+    end
+
+    pullback = Symbol(:scatter_, opname(op), :_pullback)
+    @eval function rrule(::typeof(scatter), op::typeof($op), src::AbstractArray, idx::AbstractArray)
+        y = scatter(op, src, idx)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_src(op, Δ, src, idx), DoesNotExist())
+        y, $pullback
+    end
+end
+
+for op in [max, min]
+    pullback = Symbol(:scatter!_, Symbol(op), :_pullback)
+    @eval function rrule(::typeof(scatter!), op::typeof($op), dst::AbstractArray, src::AbstractArray, idx::AbstractArray, dims)
+        y = scatter!(op, copy(dst), src, idx, dims)
+        mask_y = (dst .== y)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_dst!(op, Δ, mask_y), ∇scatter_src!(op, Δ, dst, src, idx), DoesNotExist(), DoesNotExist())
+        y, $pullback
+    end
+
+    pullback = Symbol(:scatter_, Symbol(op), :_pullback)
+    @eval function rrule(::typeof(scatter), op::typeof($op), src::AbstractArray, idx::AbstractArray)
+        y = scatter(op, src, idx)
+        $pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_src(op, Δ, y, src, idx), DoesNotExist())
+        y, $pullback
+    end
+end
+
+function rrule(::typeof(scatter!), op::typeof(mean), dst::AbstractArray, src::AbstractArray{Tsrc,Nsrc}, idx::AbstractArray{Tidx,Nidx}) where {Tsrc,Tidx,Nsrc,Nidx}
+    dims = Nsrc - Nidx
+    y = scatter!(op, copy(dst), src, idx)
+    scatter!_mean_pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_dst!(op, Δ), ∇scatter_src!(op, Δ, idx, dims), DoesNotExist())
+    y, scatter!_mean_pullback
+end
+
+function rrule(::typeof(scatter), op::typeof(mean), src::AbstractArray{Tsrc,Nsrc}, idx::AbstractArray{Tidx,Nidx}) where {Tsrc,Tidx,Nsrc,Nidx}
+    dims = Nsrc - Nidx
+    y = scatter(op, src, idx)
+    scatter_mean_pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_src(op, Δ, idx, dims), DoesNotExist())
+    y, scatter_mean_pullback
+end
```

Comment:
What's the need of specializing all rules on the op type? Can just do

```julia
function rrule(::typeof(scatter), op, src::AbstractArray{Tsrc,Nsrc}, idx::AbstractArray{Tidx,Nidx}) where {Tsrc,Tidx,Nsrc,Nidx}
    dims = Nsrc - Nidx
    y = scatter(op, src, idx)
    scatter_pullback(Δ) = (NO_FIELDS, NO_FIELDS, ∇scatter_src(op, Δ, idx, dims), DoesNotExist())
    y, scatter_pullback
end
```

Reply:
OK.

Reply:
I mean, not just the names, but removing the `for op in .... @eval`  and just define the generic method i wrote above

Reply:
You want a generic function for all?
Well............it is feasible but it will introduce several redundant arguments. Is that all right?

Reply:
mhhmh, probably capturing all would prevent some variables not needed in the backward pass from being garbage collected. Ok, let's leave things this way, we can revisit later if needed.